### PR TITLE
Fix clang-tidy support for tests and extensibility

### LINKED ISSUE #236: Ensure test suite builds with clang-tidy
The formatter action should ensure that the test-suite builds with clang-tidy enabled:
```bash
cmake -BBuild -DENABLE_CLANG_TIDY=ON -DCLANG_TIDY_BINARY=clang-tidy-11 -DBUILD_EXAMPLES=ON -DBUILD_TESTS=on .
```

https://github.com/PennyLaneAI/pennylane-lightning/blob/2cb27272c7882c5cf63a0ef047dc37884c113cac/.github/workflows/format.yml#L57

### PR DESCRIPTION
**Pull request by mlxd**
**Context:** This closes #236 and #235 

**Description of the Change:** Clang-tidy is enabled for CI tests in the formatter action, and subsequent clang-tidy suggestions are applied. Also, The return type for data is modified to become auto-definable at compile time.

**Benefits:**

**Possible Drawbacks:**

**Related GitHub Issues:**

### CODE REVIEW

**Comment by mlxd** [comment]
> Hi @mlxd, it seems like the current error is due to several unused parameters/variables. In the previous version, I have just suppressed those warnings (only for tests codes) by `-Wno-unused` as there are too many places with this warning and I thought they are not much harmful (even though some variables are unused, they provide some information e.g. `const auto num_qubits = 3;`). See the lines below.
> 
> https://github.com/PennyLaneAI/pennylane-lightning/blob/2cb27272c7882c5cf63a0ef047dc37884c113cac/pennylane_lightning/src/tests/CMakeLists.txt#L43-L45

Thanks @chaeyeunpark . I have added the casts to get over the failures for now. As discussed, we can look at adapting this to another structure later on.

File: pennylane_lightning/src/tests/Test_DynamicDispatcher.cpp [hunk 1]
```diff
@@ -68,7 +68,12 @@ struct testDispatchForKernel {
                              GateImplementation::implemented_gates, gate_op),
                          bool> = true>
     static void test(RandomEngine &re, size_t num_qubits) {
-    } // Do nothing if not implemented
+        // Keep source, but allow clang-tidy to pass for unused
+        static_cast<void>(re);
+        static_cast<void>(num_qubits);
+    } // Do nothing if not implemented;
+      // This could probably be replaced with an enable_if or SFINAE-like
+      // pattern.
```
Comment (maliasadi) [inline pennylane_lightning/src/tests/Test_DynamicDispatcher.cpp:76]: :100:

**Review by maliasadi — APPROVED** [review]
Thanks @mlxd for fixing these issues.

**Review by chaeyeunpark — CHANGES_REQUESTED** [review]
Hi @mlxd, thanks for dealing with this problem. Could you just check one thing here? If `clang-tidy` still complains about it, let's just remove this whole function (it anyway does nothing now...).

File: pennylane_lightning/src/tests/Test_OpToMemberFuncPtr.cpp [hunk 2]
```diff
@@ -22,13 +22,6 @@ template <class PrecisionT, class ParamT, class GateImplemenation,
           uint32_t gate_idx>
 constexpr bool testAllGatesImplementedIter() {
     if constexpr (gate_idx < static_cast<uint32_t>(GateOperation::END)) {
-        constexpr auto gate_op = static_cast<GateOperation>(gate_idx);
-        if constexpr (gate_op != GateOperation::Matrix) {
-            if (GateOpToMemberFuncPtr<PrecisionT, ParamT, GateImplemenation,
-                                      gate_op>::value == nullptr) {
-                return false;
-            }
-        }
```
[the remark this answers is not in the document]
Comment (mlxd) [reply pennylane_lightning/src/tests/Test_OpToMemberFuncPtr.cpp:31]: Hey @chaeyeunpark 

Ah, I understand your intention now. Yea, I was getting errors with these locally, and removed them based on the clang-tidy feedback. I'll re-add and retest with the `act` package, and if so we can leave them as deleted for now.

[the remark this answers is not in the document]
Comment (mlxd) [reply pennylane_lightning/src/tests/Test_OpToMemberFuncPtr.cpp:31]: Huh, seems to be passing now (weird, may have been a version issue locally). I'll re-add this in and push it back up. Sorry for the confusion.

Update: scratch that, fails locally on clang-tidy-13

[the remark this answers is not in the document]
Comment (mlxd) [reply pennylane_lightning/src/tests/Test_OpToMemberFuncPtr.cpp:31]: I will do some tests with different environments to see if this is an easy mitigation.

[the remark this answers is not in the document]
Comment (chaeyeunpark) [reply pennylane_lightning/src/tests/Test_OpToMemberFuncPtr.cpp:31]: As it cannot be `nullptr` anyway, I think replacing the internal if statement just to 
```
static_cast<void>(GateOpToMemberFuncPtr<PrecisionT, ParamT, GateImplemenation, gate_op>::value);
```
will work.

[the remark this answers is not in the document]
Comment (mlxd) [reply pennylane_lightning/src/tests/Test_OpToMemberFuncPtr.cpp:31]: Hey @chaeyeunpark sorry for taking so long to get back to this. That above suggestion does indeed fix the clang-tidy-13 complaint! Thanks for the suggestion.

[the remark this answers is not in the document]
Comment (chaeyeunpark) [reply pennylane_lightning/src/tests/Test_OpToMemberFuncPtr.cpp:31]: Good!

**Review by chaeyeunpark — APPROVED** [review]
Happy to approve this :+1:

**Comment by mlxd** [comment]
Thanks @chaeyeunpark & @maliasadi

**Pull request merged**